README

Tripletail
==========

Tripletail is an open-source framework which is for web applications
treat Japanese. It is supposed to run on Linux + Apache + MySQL + Perl.


INSTALLATION
~~~~~~~~~~~~
To install this module, run the following commands:

    perl Makefile.PL
    make
    make test
    make install


DEPENDENCIES
~~~~~~~~~~~~
requires:
    Crypt::CBC
    Crypt::Rijndael
    Data::Dumper
    DBI
    Digest::HMAC_SHA1
    File::Path
    IO::Scalar
    MIME::Tools                    5.411
    PadWalker                      1.0
    Test::Exception
    Test::More
    Text::CSV_XS
    Unicode::Japanese


SEE ALSO
~~~~~~~~

For further information, please refer to http://tripletail.jp/


COPYRIGHT AND LICENSE
~~~~~~~~~~~~~~~~~~~~~

Copyright (C) 2006 YMIRLINK Inc.

This framework is free software; you can redistribute it and/or modify
it under the same terms as Perl itself



Changes
Revision history for Tripletail

0.46 Tue May 26 16:56:32 JST 2009
	- FastCGIで長時間アクセスがない場合に，MySQLとのDB接続が
	　タイムアウトして切れていると，その後のアクセスでエラーが
	　出てしまう．この問題を回避するため，リクエストごとに
	　DB接続状態を確認し，必要に応じて再接続するように修正．
	- LWPの依存をMakefile.PLに追記．
	- pagerのテストの実行に失敗するバグを修正．
	- 出力フィルタとしてMobileHTMLを使用したとき、パラメータにtype => 'xhtml'を指定すると
	  コンテントタイプがapplication/xhtml+xmlになるように変更.
	- $TL->dispatchを使用したときにプロセス名に分岐先のコマンドを追加するようにした.
	  本機能を新しい INI パラメータ [TL]command_add_processnameで使用するかを選択できるようにした.

0.45
	- Validator MultiValues で 0 件は受理するように修正.
	- $TL->eval() を追加
	- [[重要:非互換な仕様変更]]
	　- $TL->escapeJS 及び $TL->unescapeJS が"<"及び">"をそれぞれ"\x3c"及び
	　  "\x3e" と変換するように変更.
	　- Form#getFileName() はベース名部分のみを返すように変更.
	- Value#isPassword() でパラメータ指定をできるように変更.
	- Pager#setToLink() メソッドの追加.

0.44 Tue Sep 16 16:27:52 JST 2008
	- outputbuffering 使用時に自動で追加されるContent-Lengthが, ユーザが任意のヘッダを追加してる場合に狂うバグの修正.
	- TL#dispatchで、第1引数が小文字ではじまる場合にエラーになるケースで、
	　　onerrorが設定されている場合は、onerrorが呼び出されるが、その後に第1引数によっては関数が実行されてしまう事があったバグの修正。
	- $TL->dispatch() に args パラメータを追加。
	- FCGI時のフィルタの再初期化をresetメソッドに分離(内部メソッド_resetもリネーム).
	- Value#isChar メソッドの追加.
	- 複値のselectへのsetFormを改善.

0.43 Wed Jun 11 15:50:54 JST 2008
	- outputbuffering 使用時に自動で追加されるContent-Lengthの計算が間違うことがあるバグの修正.

0.42 Thu Jun  5 11:27:13 JST 2008
	- Template で, includeされるファイルの文字コードが常に自動認識だったのを, 親で指定されていればそれと同じ文字コードを使うように修正.
	- Template.rootpath が相対パスで記述されている時に全てが不許可となるのを修正.
	- DBで接続用グループが存在しなかったときのエラーメッセージを加筆.
	- stratCgi外でのエラー(含コンパイルエラー)がInternal Server Errorとなっていたのをエラーメッセージを表示するように修正(compat_no_trap_for_cgi_internal_error).
	- Errorオブジェクトに http_status_line フィールドを追加.
	- $TL->printでのエラーがデバッグスクリーンにならないバグの修正.
	- Value#isPcPortable でAUの絵文字が正しく判定されない問題を修正．
	- enable_debugした際に出力が常にバッファリングされていたバグの修正.

0.41 Mon Jan 28 12:27:55 JST 2008
	- Makefile.PL がパラメータによって失敗することがあるのを修正.

0.40 Fri Jan 25 10:59:21 JST 2008
	- allow_mutable_input_cgi_object パラメータを追加.
	- 新しい INI パラメータ [TL]fcgilog
	- 新しい INI パラメータ [TL]filelog

0.39 Thu Nov 29 14:03:48 JST 2007
	- gmake 変数を使っていた箇所を除去.
	- 環境によってhtml生成に失敗する不具合を修正.

0.38 Fri Nov 23 12:15:01 JST 2007
	- [[重要:非互換な仕様変更]]
	　- $TL->CGI 及び $CGI は const 状態になっていなかったが、const 状態にするよ
	　　うにした。
	　- TL#dispatchで、第1引数が小文字ではじまる場合にエラーになるケースで、
	　　onerrorが設定されている場合は、onerrorが呼び出されるようにした。
	- Value#isDomainName() が正当でないドメイン名に対して 1 を返す場合があった
	　問題を解決。
	- Value#isDateTime() を追加。
	- TL#writeTextFileの文字コード変換が正常に動作していなかった問題の修正。
	- [TL]outputbuffering が 1 でない場合、$TL->print 等で何らかのデータを出力し
	　た後にエラーが発生した時に表示が崩れる問題を解決。
	- 0.35で導入された、Template#getForm, setForm, extForm, addHiddenForm,
	　addSessionCheck において対象となるノードに未置換の <&タグ> がある場合、
	　エラーになるように変更された修正を回避するオプションを追加した。
	　互換性維持の為に使用されるオプションであり、非推奨オプションとなります。
	- const 状態の Form オブジェクトを変更可能にするオプションを [TL] の ini に追
	　加した。

0.37 Thu Oct  4 18:09:30 JST 2007
	- Win32 で動かなくなっていた箇所の修正.
	- IIS/FastCGI/ActivePerlの対応.
	- Value に isDomainName() を追加。Validator のフィルタに DomainName を追加。
	　いずれもドメイン名として正当であるかどうかを確認する。
	- Value に detectMobileAgent() を追加。User-Agent 名から文字コード名を判別す
	　る。
	- Pager#setFormParam の引数が hashref をとることも出来るように修正した。
	- Tripletail::Sendmail::Smtp でエラーが発生した場合、相手サーバからの
	　最後の応答をエラーメッセージに含めるように修正した。

0.36 Fri Sep 28 11:03:31 JST 2007
	- [[重要:非互換な仕様変更]]
	　- Template#getForm, setForm, extForm, addHiddenForm, addSessionCheck は対象
	　　となるノードに単に未置換の <&タグ> があるだけではエラーにならないように変
	　　更された。<foo <&tag>> のように HTML タグの内部に未置換のテンプレートタグ
	　　がある場合にはエラーになる。<foo <!begin:bar>> 等も同様にエラーになる。

0.35 Thu Sep 27 12:03:15 JST 2007
	- [[重要:非互換な仕様変更]]
	　- Template#getForm, setForm, extForm, addHiddenForm, addSessionCheck は対象
	　　となるノードに未置換の <&タグ> がある場合にはエラーになるように変更された。
	　- DateTime はこれまで YYYY-M-D H:M:S のように月日時分秒が一桁である形式の日
	　　付を受け入れなかったが、受け入れるようになった。但し区切り記号が存在する
	　　場合のみ。
	- DBのクエリをデバッグする際に、パラメータが記録に残らないバグを修正。
	- 意味の解り難い英文のエラーメッセージを全面的に見直し。
	- Tripletail::SMIME を Tripletail フレームワークから独立させ、Crypt::SMIME に
	　した。$TL->newSMIME は継続して利用可能。
	- $TL->getCookie と $TL->getRawCookie は $TL->startCgi の外から呼ぶとエラーに
	　なるようにした。
	- DB のクエリログを取っている場合、これまでは $TL->startCgi() の内部で実行さ
	　れたクエリであるかどうかに関係無くデバッグ表示画面に表示する為に内部に全て
	　のクエリログを保存していたが、$TL->startCgi() 内部で実行されたものでなけれ
	　ば保存する意味が無いため、保存しないようにした。
	- FastCGI 環境で動作しているかどうかを判定する方法を、FastCGI の仕様書に定め
	　られている通りにした。
	- $TL->startCgi() の -main 関数の外でエラーが発生した場合にそのエラー内容が二
	　重にログに出力される問題を解決。
    - Pager の必須ノードであった PageNumLinks は必須でなくなった。

0.34 Fri Sep 14 14:38:19 JST 2007
	- [[重要:非互換な仕様変更]]
	　- Value#isHira, isKata で判定をする際、「ー」があると偽を
	　　返していたが、真を返すように修正。
	　- Tripletail::Filter::CSV に配列のリファレンスを渡した場合に、
	　　出力時の改行コードを "\n" から "\r\n" に変更した。
	　- startCgi内で、不完全なリクエスト
	　　(POST時にbodyがContent-Lengthよりも短いもの等)
	　　の場合は、途中で接続が切れたものとして扱い、
	　　その時はinitRequestのみ実行され、
	　　pre/postRequestフック及びmainは実行されずに、
	　　ダミーのコンテンツを返すようにした
	　　(これはエラー扱いではないので、エラーログ及びエラー通知メールは
	　　 送られない)
	　- Tripletail::Filter::MemCachedのmodeオプションのreadが、
	　　pass-throughに変更されました。
	　- Tripletail#printCacheUnlessModified, Tripletail#setCacheの、
	　　引数を変更し、Tripletaik#setCacheFilterを追加した。
	- errortemplateを指定した場合に正常に機能しない問題を修正した。
	- Tripletail::Filter::CSV に配列のリファレンスを渡した場合に
	　どのような改行コードを利用するか指定するオプションを追加。

0.33 Mon Sep 10 16:26:55 JST 2007
	- [[重要:非互換な仕様変更]]
	　- DateTimeクラスのtai64フォーマットの処理に問題があり、修正が
	　　難しいため、tai64フォーマットへの対応を削除。
	　- Tripletail::Filter::MemCachedのtypeオプションはmodeオプションに
	　　名称を変更し、引数はin/outからwrite/readに変更されました。
	- Pagerクラスにて、DBセットを利用する手段が無かった為、
	　Pager#newPager時にdb_objectを渡せるように変更しました。
	　今後はこちらの形式が推奨方式になります。
	　db_groupを渡す方法も継続して利用できますが、非推奨な方法となりました。
	　あわせて、Pager#setDbGroupも非推奨となりました。
	- ValueクラスのforcePortable、forcePcPortableの記述間違いを修正。
	　正しい動作は機種依存文字を削除する。
	- SessionクラスがSQLの実行に失敗した時の処理が適切でなかったので修正。
	- $TL->sendError のマニュアルのサンプルの記述ミスを修正。
	- t/value.t が希に失敗することがある問題を修正。
	- taint 時に動かない箇所の修正.
	- TLのoutputbufferingを1にすると2重にフィルタがかかるバグを修正。
	- TLのoutputbufferingが1の場合は、Content-Lengthヘッダを付与するように修正。
	- Filter::MobileHTMLクラス利用時に、Content-Lengthヘッダを返すように
	　outputbufferingを強制的に1にセットするように修正。
	　（携帯向けにコンテンツを返す場合、Content-Lengthヘッダを必須とする
	　　キャリアがあるため）
	- errortemplateを指定した場合、そのテンプレートに対してデバッグ出力などの
	　ノードを展開しようとしていた問題を修正。指定したテンプレートはそのまま
	　出力するように修正された。
	　また、errortemplatecharset の設定を追加した。
	- Windows環境でmail.tのテストに失敗する問題を(たぶん)修正。
	- 特定の環境でsmime.tのテストに失敗する問題を修正。

Makefile.PL
use strict;
use warnings;
use ExtUtils::MakeMaker;
use inc::ExtUtils::MY_Metafile;

my $requires = {
    'DBI'                 => 0,
    'Data::Dumper'        => 0,
    'Digest::HMAC_SHA1'   => 0,
    'Encode'              => 0,
    'File::Path'          => 0,
    'IO::Scalar'          => 0, # IO-stringy
    'MIME::Tools'         => '5.411',
    'PadWalker'           => '1.0',
    'Unicode::Japanese'   => '0.43',
    'Text::CSV_XS'        => 0, # optional
    'LWP'                 => 0,
};

my $build_requires = {
    'Crypt::CBC'          => 0,
    'Crypt::Rijndael'     => 0,
    'IO::ScalarArray'     => '2.110', # IO-stringy, older one has inf-loop bug?
    'Test::Exception'     => 0,
    'Test::More'          => 0,
};

my_metafile Tripletail => {
  no_index => {
    directory => [qw(sample t)],
  },
  requires => $requires,
  build_requires => $build_requires,
  license => 'perl',
};

WriteMakefile(
    NAME                => 'Tripletail',
    AUTHOR              => 'Ymirlink <tl@tripletail.jp>',
    VERSION_FROM        => 'lib/Tripletail.pm',
    ABSTRACT_FROM       => 'lib/Tripletail.pm',
    PL_FILES            => {},
    EXE_FILES           => [qw(bin/tltail.pl)],
    PREREQ_PM => {
        %$build_requires,
        %$requires,
    },
    dist                => { COMPRESS => 'gzip -9f', SUFFIX => 'gz', },
);

# -----------------------------------------------------------------------------
# End of File.
# -----------------------------------------------------------------------------
